File: backend/app/ai/anomaly_detector.py
"""Voxel-cluster anomaly detection on top of the fused probability volume."""
from __future__ import annotations

from dataclasses import dataclass, field

import numpy as np
from scipy import ndimage

from app.ai.classifier import classify_evidence
from app.core.materials import MATERIALS
from app.core.voxel_engine import VoxelGrid
from app.sensors.base import SensorSimulationResult

_CONNECTIVITY_26 = np.ones((3, 3, 3), dtype=int)


@dataclass
class DetectedAnomaly:
    id: int
    centroid_vox: tuple[float, float, float]
    centroid_m: tuple[float, float, float]
    size_voxels: int
    volume_m3: float
    bbox_vox: tuple[tuple[int, int], tuple[int, int], tuple[int, int]]
    fused_confidence: float
    agreement_count: int
    contributing_sensors: list[str]
    predicted_material: str
    predicted_material_name_tr: str
    classification_confidence: float
    signature: dict[str, float] = field(default_factory=dict)
    material_scores: dict[str, float] = field(default_factory=dict)


def detect_anomalies(
    grid: VoxelGrid,
    sensor_results: dict[str, SensorSimulationResult],
    fused_probability: np.ndarray,
    agreement_count: np.ndarray,
    min_size_voxels: int = 2,
    threshold_percentile: float = 92.0,
) -> list[DetectedAnomaly]:
    threshold = max(
        float(np.percentile(fused_probability, threshold_percentile)),
        float(fused_probability.mean() + fused_probability.std()),
    )
    mask = fused_probability >= threshold
    labeled, num_components = ndimage.label(mask, structure=_CONNECTIVITY_26)

    anomalies: list[DetectedAnomaly] = []
    for label_id in range(1, num_components + 1):
        component_mask = labeled == label_id
        size = int(component_mask.sum())
        if size < min_size_voxels:
            continue

        coords = np.argwhere(component_mask)
        centroid_vox = coords.mean(axis=0)
        bbox = tuple((int(coords[:, i].min()), int(coords[:, i].max())) for i in range(3))
        fused_conf = float(fused_probability[component_mask].mean())
        agreement = int(round(float(agreement_count[component_mask].mean())))

        evidence_vector: dict[str, float] = {}
        contributing: list[str] = []
        for key, result in sensor_results.items():
            value = float(result.probability_volume[component_mask].mean())
            evidence_vector[key] = value
            if value > 0.5:
                contributing.append(key)

        predicted_material, class_conf, scores = classify_evidence(evidence_vector)
        material = MATERIALS[predicted_material]

        anomalies.append(
            DetectedAnomaly(
                id=0,
                centroid_vox=tuple(float(c) for c in centroid_vox),
                centroid_m=tuple(float(c) * grid.voxel_size_m for c in centroid_vox),
                size_voxels=size,
                volume_m3=size * grid.voxel_size_m**3,
                bbox_vox=bbox,
                fused_confidence=fused_conf,
                agreement_count=agreement,
                contributing_sensors=contributing,
                predicted_material=predicted_material,
                predicted_material_name_tr=material.name_tr,
                classification_confidence=class_conf,
                signature=evidence_vector,
                material_scores=scores,
            )
        )

    anomalies.sort(key=lambda a: a.fused_confidence, reverse=True)
    for index, anomaly in enumerate(anomalies):
        anomaly.id = index + 1
    return anomalies

File: frontend/src/sim/ai/anomalyDetector.ts
import type { VoxelGrid } from "../voxelEngine";
import type { SensorSimulationResult } from "../sensors/base";
import { classifyEvidence } from "./classifier";
import { MATERIALS } from "../materials";

export interface DetectedAnomaly {
  id: number;
  centroidVox: [number, number, number];
  centroidM: [number, number, number];
  sizeVoxels: number;
  volumeM3: number;
  bboxVox: [[number, number], [number, number], [number, number]];
  fusedConfidence: number;
  agreementCount: number;
  contributingSensors: string[];
  predictedMaterial: string;
  predictedMaterialNameTr: string;
  classificationConfidence: number;
  signature: Record<string, number>;
  materialScores: Record<string, number>;
}

function percentile(sorted: Float64Array, p: number): number {
  const idx = (p / 100) * (sorted.length - 1);
  const lo = Math.floor(idx);
  const hi = Math.ceil(idx);
  if (lo === hi) return sorted[lo];
  return sorted[lo] + (sorted[hi] - sorted[lo]) * (idx - lo);
}

const NEIGHBOR_OFFSETS: [number, number, number][] = [];
for (let dx = -1; dx <= 1; dx++)
  for (let dy = -1; dy <= 1; dy++)
    for (let dz = -1; dz <= 1; dz++) if (dx !== 0 || dy !== 0 || dz !== 0) NEIGHBOR_OFFSETS.push([dx, dy, dz]);

export function detectAnomalies(
  grid: VoxelGrid,
  sensorResults: Record<string, SensorSimulationResult>,
  fusedProbability: Float64Array,
  agreementCount: Int32Array,
  minSizeVoxels = 2,
  thresholdPercentile = 92.0
): DetectedAnomaly[] {
  const sorted = fusedProbability.slice().sort();
  let mean = 0;
  for (const v of fusedProbability) mean += v;
  mean /= fusedProbability.length;
  let variance = 0;
  for (const v of fusedProbability) variance += (v - mean) ** 2;
  const std = Math.sqrt(variance / fusedProbability.length);
  const threshold = Math.max(percentile(sorted, thresholdPercentile), mean + std);

  const { nx, ny, nz } = grid;
  const visited = new Uint8Array(nx * ny * nz);
  const anomalies: DetectedAnomaly[] = [];
  const sensorKeys = Object.keys(sensorResults);

  for (let x0 = 0; x0 < nx; x0++) {
    for (let y0 = 0; y0 < ny; y0++) {
      for (let z0 = 0; z0 < nz; z0++) {
        const start = grid.idx(x0, y0, z0);
        if (visited[start] || fusedProbability[start] < threshold) continue;

        const component: number[] = [];
        const stack: [number, number, number][] = [[x0, y0, z0]];
        visited[start] = 1;
        while (stack.length > 0) {
          const [x, y, z] = stack.pop()!;
          component.push(grid.idx(x, y, z));
          for (const [dx, dy, dz] of NEIGHBOR_OFFSETS) {
            const nx2 = x + dx;
            const ny2 = y + dy;
            const nz2 = z + dz;
            if (nx2 < 0 || nx2 >= nx || ny2 < 0 || ny2 >= ny || nz2 < 0 || nz2 >= nz) continue;
            const ni = grid.idx(nx2, ny2, nz2);
            if (visited[ni] || fusedProbability[ni] < threshold) continue;
            visited[ni] = 1;
            stack.push([nx2, ny2, nz2]);
          }
        }

        if (component.length < minSizeVoxels) continue;

        let sx = 0, sy = 0, sz = 0;
        let xMin = Infinity, xMax = -Infinity, yMin = Infinity, yMax = -Infinity, zMin = Infinity, zMax = -Infinity;
        let fusedSum = 0;
        let agreementSum = 0;
        for (const i of component) {
          const x = Math.floor(i / (ny * nz));
          const y = Math.floor((i % (ny * nz)) / nz);
          const z = i % nz;
          sx += x; sy += y; sz += z;
          xMin = Math.min(xMin, x); xMax = Math.max(xMax, x);
          yMin = Math.min(yMin, y); yMax = Math.max(yMax, y);
          zMin = Math.min(zMin, z); zMax = Math.max(zMax, z);
          fusedSum += fusedProbability[i];
          agreementSum += agreementCount[i];
        }
        const n = component.length;
        const centroidVox: [number, number, number] = [sx / n, sy / n, sz / n];

        const evidenceVector: Record<string, number> = {};
        const contributing: string[] = [];
        for (const key of sensorKeys) {
          const vol = sensorResults[key].probabilityVolume;
          let sum = 0;
          for (const i of component) sum += vol[i];
          const value = sum / n;
          evidenceVector[key] = value;
          if (value > 0.5) contributing.push(key);
        }

        const { predicted, confidence, scores } = classifyEvidence(evidenceVector);
        const material = MATERIALS[predicted];

        anomalies.push({
          id: 0,
          centroidVox,
          centroidM: [centroidVox[0] * grid.voxelSizeM, centroidVox[1] * grid.voxelSizeM, centroidVox[2] * grid.voxelSizeM],
          sizeVoxels: n,
          volumeM3: n * grid.voxelSizeM ** 3,
          bboxVox: [[xMin, xMax], [yMin, yMax], [zMin, zMax]],
          fusedConfidence: fusedSum / n,
          agreementCount: Math.round(agreementSum / n),
          contributingSensors: contributing,
          predictedMaterial: predicted,
          predictedMaterialNameTr: material.name_tr,
          classificationConfidence: confidence,
          signature: evidenceVector,
          materialScores: scores,
        });
      }
    }
  }

  anomalies.sort((a, b) => b.fusedConfidence - a.fusedConfidence);
  anomalies.forEach((a, i) => (a.id = i + 1));
  return anomalies;
}
